src/Plugin/authentication/OAuth2ServerAuthentication.php

```diff
@@ -51,8 +51,22 @@ public function authenticate(RequestInterface $request) {
     if ($result instanceof \OAuth2\Response) {
       throw new UnauthorizedException($result->getResponseBody(), $result->getStatusCode());
     }
-    elseif (empty($result['user_id'])) {
-      return NULL;
+
+    if (empty($result['user_id'])) {
+      // If the user_id is not set, it could mean that this is a client
+      // credentials grant token, in which case the client_id would be set.
+      if (empty($result['client_id'])) {
+        return NULL;
+      }
+
+      // We are dealing with client credentials flow. See if the resource has
+      // defined an user for this grant type.
+      if (!empty($oauth2_info['client_credentials_uid'])) {
+        $result['user_id'] = $oauth2_info['client_credentials_uid'];
+      }
+      elseif (!empty($oauth2_info['client_credentials_user'])) {
+        $result['user_id'] = user_load_by_name($oauth2_info['client_credentials_user'])->uid;
```

Comment:
We are loading the user by name and then by id. We should load the user only once.